synapseclient/core/otel_config.py

```diff
@@ -0,0 +1,234 @@
+"""OpenTelemetry configuration for Synapse Python Client."""
+import os
+import sys
+from typing import Any, Dict, List, Optional
+
+from opentelemetry import metrics, trace
+from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.metrics.export import (
+    ConsoleMetricExporter,
+    PeriodicExportingMetricReader,
+)
+from opentelemetry.sdk.resources import (
+    SERVICE_INSTANCE_ID,
+    SERVICE_NAME,
+    SERVICE_VERSION,
+    Resource,
+)
+from opentelemetry.sdk.trace import SpanProcessor, TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
+from opentelemetry.trace import Span, SpanContext, get_current_span
+
+# Default service name for the client
+DEFAULT_SERVICE_NAME = "synapseclient"
+CLIENT_VERSION = "unspecified"
+SYNAPSE_SERVICE_VERSION = f"synapse.{SERVICE_VERSION}"
+
+
+class AttributePropagatingSpanProcessor(SpanProcessor):
```

Comment:
From what I remember, we have a lot of more complicated context managers set throughout the code when we first enabled otel across our functions.  Will this improvement reduce the number of those?

Reply:
The code I added into client.py and fhe otel_config.py is meant to simplify some usage. I added to the Readme and in the docstrings to show how they can be used. That combined with the auto instrumentation libraries is meant to make it easier to use.